cpp/ql/src/semmle/code/cpp/rangeanalysis/Bound.qll

```diff
@@ -8,8 +8,8 @@ private newtype TBound =
     exists(Instruction i |
       vn.getAnInstruction() = i and
       (
-        i.getResultType() instanceof IntegralType or
-        i.getResultType() instanceof PointerType
+        i.getResultIRType() instanceof IRIntegerType or
+        i.getResultIRType() instanceof IRAddressType
```

Comment:
I was just trying to use the range analysis library to validate an index into a stack-allocated array, and I found that it didn't work. After a long debugging session, I figured out that it was because this predicate and `SafeCastInstruction` didn't treat a glvalue as a pointer, so a local variable subjected to array-to-pointer decay would not become a `Bound`. After I'd fixed that, I remembered this PR and came back here to discover that this innocent-looking change also has exactly that effect! So that's good.

The question is now what it means for performance that all variable addresses are now `Bound`s. @rdmarsh2, what do you think about these changes to the range analysis library? @dave-bartolomeo, please check how performance and size of `boundedInstruction` change.

Reply:
For local variables with an array type or a struct type with an array member, that should be an improvement. I'm a bit nervous about the performance implications of all variable addresses being bounds. It should have a bigger impact on the size of `boundedInstruction` than on the time to compute the stage, but that means the performance hit will come from users of the library rather than the library itself.